source/guides/recipes/autoformatting.md

```diff
@@ -0,0 +1,44 @@
+(autoformatting)=
+# Autoformatting
+
+[`nixfmt`](https://github.com/NixOS/nixfmt) is the official Nix autoformatter.
+
+It is not yet wired up to any official tooling. Subscribe to [PR
+11252 to NixOS/nix](https://github.com/NixOS/nix/pull/11252) for updates on
+that effort.
+
+The easiest way to leverage `nixfmt` in your project is with the `nixfmt-tree`
+wrapper. Here's how:
+
+- For `nix fmt` to format all Nix files, add this to the `flake.nix` outputs:
+
+  ```nix
+  formatter.''${system} = nixpkgs.legacyPackages.''${system}.nixfmt-tree;
+  ```
+
+- The same can be done more efficiently with the `treefmt` command,
+  which you can get in `nix-shell`/`nix develop` by extending `mkShell` using
+
+  ```nix
+  mkShell {
+    packages = [ pkgs.nixfmt-tree ];
+  }
+  ```
+
+  You can then also use `treefmt` in a pre-commit/pre-push [Git hook](https://git-scm.com/docs/githooks)
+  and `nixfmt` with your editors format-on-save feature.
+
+- To check formatting in CI, run the following in a checkout of your Git repository:
+  ```
+  treefmt --ci
+  ```
+
+For more flexibility, you can customize this package using
+```nix
+nixfmt-tree.override {
+  settings = { /* treefmt config */ };
+  runtimePackages = [ /* List any formatters here */ ];
+}
+```
+
+Alternatively you can switch to the more fully-featured [treefmt-nix](https://github.com/numtide/treefmt-nix).
```

Comment:
Rather than propose ever more alternative ways of doing things, I'd appreciate (and review) a guide to set up a pre-commit hook with treefmt-nix.

Reply:
While they both relate to formatting, and both produce a configured instance of treefmt, `pkgs.nixfmt-tree`, `pkgs.treefmt.withConfig`, and `treefmt-nix` all exist for different reasons and suit different use cases.

Users who _just_ want to configure formatting nix files in their project shouldn't need to download or configure treefmt-nix; there's a solution built into nixpkgs.

Users who want a more complex formatting setup, where they end up configuring several treefmt options, may benefit from using either treefmt-nix or `pkgs.treefmt.withConfig`.

Reply:
Right, and then we'd need to ask what's the purpose of any given article, and whether this information may be better placed elsewhere, such as is Nixpkgs contributor docs.

This sort of problem appears all of the time here, and sometimes one first needs to write an entirely guide to to write the guide one originally wanted to write. Otherwise it becomes just as unfocused and confusing as we still have an unfortunate reputation for.